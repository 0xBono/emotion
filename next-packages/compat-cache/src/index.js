// @flow

let createCompatCache = (emotion: *) => {
  // rename this property to cache later and remove this package and tell people to use .cache
  return emotion.caches
}

export default createCompatCache
