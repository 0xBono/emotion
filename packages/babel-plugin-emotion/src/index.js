// @flow
import { createEmotionMacro } from './macro'
import { createStyledMacro } from './styled-macro'
import cssMacro, { transformCssCallExpression } from './css-macro'
import { addDefault } from '@babel/helper-module-imports'
import { getSourceMap } from '@emotion/babel-utils'
import { buildStyledOptions } from './babel-utils'

let webStyledMacro = createStyledMacro('@emotion/styled-base', true)
let nativeStyledMacro = createStyledMacro('@emotion/native', false)
let primitivesStyledMacro = createStyledMacro('@emotion/primitives', false)

export const macros = {
  emotion: createEmotionMacro('emotion'),
  css: cssMacro,
  styled: webStyledMacro
}

export type BabelPath = any

export type EmotionBabelPluginPass = any

let emotionCoreMacroThatsNotARealMacro = ({ references, state, babel }) => {
  Object.keys(references).forEach(refKey => {
    if (refKey === 'css') {
      references[refKey].forEach(path => {
        transformCssCallExpression({ babel, state, path: path.parentPath })
      })
    }
  })
}
emotionCoreMacroThatsNotARealMacro.keepImport = true

let pluginMacros = {
  '@emotion/css': cssMacro,
  '@emotion/styled': webStyledMacro,
  '@emotion/core': emotionCoreMacroThatsNotARealMacro,
  'react-emotion': webStyledMacro,
  '@emotion/primitives': primitivesStyledMacro,
  '@emotion/native': nativeStyledMacro,
  emotion: createEmotionMacro('emotion')
}

export default function(babel: *) {
  let t = babel.types
  return {
    name: 'emotion',
    inherits: require('babel-plugin-syntax-jsx'),
    visitor: {
      ImportDeclaration(path: *, state: *) {
        // most of this is from https://github.com/kentcdodds/babel-plugin-macros/blob/master/src/index.js
        if (pluginMacros[path.node.source.value] === undefined) {
          return
        }
        const imports = path.node.specifiers.map(s => ({
          localName: s.local.name,
          importedName:
            s.type === 'ImportDefaultSpecifier' ? 'default' : s.imported.name
        }))
        let shouldExit = false
        let hasReferences = false
        const referencePathsByImportName = imports.reduce(
          (byName, { importedName, localName }) => {
            let binding = path.scope.getBinding(localName)
            if (!binding) {
              shouldExit = true
              return byName
            }
            byName[importedName] = binding.referencePaths
            hasReferences =
              hasReferences || Boolean(byName[importedName].length)
            return byName
          },
          {}
        )
        if (!hasReferences || shouldExit) {
          return
        }
        /**
         * Other plugins that run before babel-plugin-macros might use path.replace, where a path is
         * put into its own replacement. Apparently babel does not update the scope after such
         * an operation. As a remedy, the whole scope is traversed again with an empty "Identifier"
         * visitor - this makes the problem go away.
         *
         * See: https://github.com/kentcdodds/import-all.macro/issues/7
         */
        state.file.scope.path.traverse({
          Identifier() {}
        })

        pluginMacros[path.node.source.value]({
          references: referencePathsByImportName,
          state,
          babel,
          isBabelMacrosCall: true
        })
        if (!pluginMacros[path.node.source.value].keepImport) {
          path.remove()
        }
      },
      Program(path: *, state: *) {
        if (state.opts.jsx === undefined) {
          for (const node of path.node.body) {
            if (
              t.isImportDeclaration(node) &&
              node.source.value === '@emotion/core'
            ) {
              state.transformCssProp = true
              break
            }
          }
        } else {
          state.transformCssProp = state.opts.jsx
        }

        if (state.opts.sourceMap) {
          state.emotionSourceMap = true
        }
      },
      JSXAttribute(path: *, state: *) {
        if (path.node.name.name !== 'css' || !state.transformCssProp) {
          return
        }

        if (
          t.isJSXExpressionContainer(path.node.value) &&
          (t.isObjectExpression(path.node.value.expression) ||
            t.isArrayExpression(path.node.value.expression))
        ) {
          let expressionPath = path.get('value.expression')
          if (expressionPath.isPure()) {
            if (!state.cssIdentifier) {
              state.cssIdentifier = addDefault(path, '@emotion/css', {
                nameHint: 'css'
              })
            }
            expressionPath.replaceWith(
              t.callExpression(
                t.cloneDeep(state.cssIdentifier),
                [
                  path.node.value.expression,
                  state.emotionSourceMap &&
                    path.node.loc !== undefined &&
                    t.stringLiteral(getSourceMap(path.node.loc.start, state))
                ].filter(Boolean)
              )
            )
            transformCssCallExpression({
              babel,
              state,
              path: expressionPath
            })
          }
        }
      },
      CallExpression: {
        exit(path: BabelPath, state: EmotionBabelPluginPass) {
          try {
            if (
              path.node.callee &&
              path.node.callee.property &&
              path.node.callee.property.name === 'withComponent'
            ) {
              switch (path.node.arguments.length) {
                case 1:
                case 2: {
                  path.node.arguments[1] = buildStyledOptions(t, path, state)
                }
              }
            }
          } catch (e) {
            throw path.buildCodeFrameError(e)
          }
        }
      }
    }
  }
}
