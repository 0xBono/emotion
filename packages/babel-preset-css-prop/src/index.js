import jsx from '@babel/plugin-transform-react-jsx'
import jsxDev from '@babel/plugin-transform-react-jsx-development'
import pragmatic from '@emotion/babel-plugin-jsx-pragmatic'
import emotion from '@emotion/babel-plugin'

let pragmaName = '___EmotionJSX'

// pull out the emotion options and pass everything else to the jsx transformer
// this means if @babel/plugin-transform-react-jsx adds more options, it'll just work
// and if @emotion/babel-plugin adds more options we can add them since this lives in
// the same repo as @emotion/babel-plugin

export default (
  api,
  {
    pragma,
    sourceMap,
    autoLabel,
    labelFormat,
    importMap,
    development = false,
    ...options
  } = {}
) => {
  if (options.runtime) {
    console.warn(
      '`runtime` option has been deprecated. It still works and will continue to work in Emotion 10 but we have found out that including JSX plugin twice in the Babel configuration leads to hard to debug problems and it\'s not always obvious that some presets include it. If you want to configure `runtime: "automatic"` just replace `@emotion/babel-preset-css-prop` with `@babel/preset-react` and `@emotion/babel-plugin`. You can find out how to configure things properly here: https://emotion.sh/docs/css-prop#babel-preset'
    )
  }

  return {
    plugins: [
      options.runtime !== 'automatic' && [
        pragmatic,
        {
          export: 'jsx',
          module: '@emotion/react',
          import: pragmaName
        }
      ],
      [
        development ? jsxDev : jsx,
        {
          ...(options.runtime === 'automatic'
            ? { importSource: '@emotion/react' }
            : { pragma: pragmaName, pragmaFrag: 'React.Fragment' }),
          ...options
        }
      ],
      [
        emotion,
        {
          sourceMap,
          autoLabel,
          labelFormat,
          cssPropOptimization: true,
          importMap
        }
      ]
    ].filter(Boolean)
  }
}
