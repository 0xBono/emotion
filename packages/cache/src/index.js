// @flow
import { StyleSheet } from '@emotion/sheet'
import {
  isBrowser,
  type EmotionCache,
  type SerializedStyles
} from '@emotion/utils'
import Stylis from '@emotion/stylis'
import weakMemoize from '@emotion/weak-memoize'
// import ruleSheetPlugin from './rule-sheet'
import type { StylisPlugin } from './types'

export type PrefixOption =
  | boolean
  | ((key: string, value: string, context: 1 | 2 | 3) => boolean)

type StylisPlugins = StylisPlugin[] | StylisPlugin

export type Options = {
  nonce?: string,
  stylisPlugins?: StylisPlugins,
  prefix?: PrefixOption,
  key?: string,
  container?: HTMLElement,
  speedy?: boolean
}

let rootServerStylisCache = {}

let getServerStylisCache = weakMemoize(() => ({}))

const delimiter = '/*|*/'
const needle = delimiter + '}'

let createCache = (options?: Options): EmotionCache => {
  if (options === undefined) options = {}
  let key = options.key || 'css'
  let stylisOptions

  if (options.prefix !== undefined) {
    stylisOptions = {
      prefix: options.prefix
    }
  }

  function toSheet(block) {
    if (block) {
      currentSheet.insert(block + '}')
    }
  }

  let ruleSheet: StylisPlugin = (
    context,
    content,
    selectors,
    parents,
    line,
    column,
    length,
    ns,
    depth,
    at
  ) => {
    switch (context) {
      // property
      case 1: {
        switch (content.charCodeAt(0)) {
          case 64: {
            // @import
            if (depth === 0) {
              currentSheet.insert(content + ';')
              return ''
            }
            break
          }
          // charcode for l
          case 108: {
            // charcode for b
            // this ignores label
            if (content.charCodeAt(2) === 98) {
              return ''
            }
          }
        }
        break
      }
      // selector
      case 2: {
        if (ns === 0) return content + delimiter
        break
      }
      // at-rule
      case 3: {
        switch (ns) {
          // @font-face, @page
          case 102:
          case 112: {
            currentSheet.insert(selectors[0] + content)
            return ''
          }
          default: {
            return content + (at === 0 ? delimiter : '')
          }
        }
      }
      case -2: {
        content.split(needle).forEach(toSheet)
        return current
      }
    }
  }

  let current

  let stylis = new Stylis(stylisOptions)

  stylis.use(options.stylisPlugins)(ruleSheet)

  let serverStylis
  if (!isBrowser) {
    let removeLabel: StylisPlugin = (context, content) => {
      if (
        context === 1 &&
        // charcode for l
        content.charCodeAt(0) === 108 &&
        // charcode for b
        content.charCodeAt(2) === 98
        // this ignores label
      ) {
        return ''
      }
    }
    serverStylis = new Stylis(stylisOptions)
    serverStylis.use(removeLabel)
    let serverStylisCache = rootServerStylisCache
    if (options.stylisPlugins) {
      serverStylis.use(options.stylisPlugins)
      serverStylisCache = getServerStylisCache(options.stylisPlugins)
    }
    let actualServerStylis = serverStylis
    serverStylis = (selector, serialized) => {
      let name = serialized.name
      if (serverStylisCache[name] === undefined) {
        serverStylisCache[name] = actualServerStylis(
          selector,
          serialized.styles
        )
      }
      return serverStylisCache[name]
    }
  }

  if (process.env.NODE_ENV !== 'production') {
    // $FlowFixMe
    if (/[^a-z-]/.test(key)) {
      throw new Error(
        `Emotion key must only contain lower case alphabetical characters and - but "${key}" was passed`
      )
    }
    stylis.use((context, content, selectors) => {
      switch (context) {
        case 2: {
          for (let i = 0, len = selectors.length; len > i; i++) {
            // :last-child isn't included here since it's safe
            // because a style element will never be the last element
            let match = selectors[i].match(/:(first|nth|nth-last)-child/)
            if (match !== null) {
              console.error(
                `The pseudo class "${
                  match[0]
                }" is potentially unsafe when doing server-side rendering. Try changing it to "${
                  match[1]
                }-of-type"`
              )
            }
          }
          break
        }
      }
    })
  }
  let inserted = {}
  // $FlowFixMe
  let container: HTMLElement
  if (isBrowser) {
    container = options.container || document.head

    const nodes = document.querySelectorAll(`style[data-emotion-${key}]`)

    Array.prototype.forEach.call(nodes, (node: HTMLStyleElement) => {
      const attrib = node.getAttribute(`data-emotion-${key}`)
      // $FlowFixMe
      attrib.split(' ').forEach(id => {
        inserted[id] = true
      })
      if (node.parentNode !== container) {
        container.appendChild(node)
      }
    })
  }
  let currentSheet

  function insert(
    selector: string,
    serialized: SerializedStyles,
    sheet: StyleSheet,
    shouldCache: boolean
  ): string | void {
    let name = serialized.name
    if (isBrowser) {
      currentSheet = sheet
      if (
        process.env.NODE_ENV !== 'production' &&
        serialized.map !== undefined
      ) {
        let map = serialized.map
        currentSheet = {
          insert: (rule: string) => {
            sheet.insert(rule + map)
          }
        }
      }
      stylis(selector, serialized.styles)
      if (shouldCache) {
        cache.inserted[name] = true
      }
    } else {
      let rules = serverStylis(selector, serialized)
      if (cache.compat === undefined) {
        // in regular mode, we don't set the styles on the inserted cache
        // since we don't need to and that would be wasting memory
        // we return them so that they are rendered in a style tag
        if (shouldCache) {
          cache.inserted[name] = true
        }
        return rules
      } else {
        // in compat mode, we put the styles on the inserted cache so
        // that emotion-server can pull out the styles
        if (shouldCache) {
          cache.inserted[name] = rules
        }
      }
    }
  }

  const cache: EmotionCache = {
    key,
    sheet: new StyleSheet({
      key,
      container,
      nonce: options.nonce,
      speedy: options.speedy
    }),
    nonce: options.nonce,
    inserted,
    registered: {},
    insert
  }
  return cache
}

export default createCache
