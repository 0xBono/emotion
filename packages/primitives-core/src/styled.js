// @flow
import * as React from 'react'
import {
  channel,
  contextTypes,
  setTheme,
  testAlwaysTrue,
  pickAssign,
  interleave
} from './utils'
import { withCSSContext } from '@emotion/core'
import { createCss } from './css'

let defaultPickTest = prop => prop !== 'theme' && prop !== 'innerRef'

type options = {
  getShouldForwardProp: (cmp: React.ElementType) => (prop: string) => boolean
}

export function createStyled(
  StyleSheet: Object,
  { getShouldForwardProp = () => defaultPickTest }: options = {}
) {
  const css = createCss(StyleSheet)

  return function createEmotion(component: React.ElementType) {
    let pickTest = getShouldForwardProp(component)

    return function createStyledComponent(...rawStyles: *) {
      let styles

      if (rawStyles[0] == null || rawStyles[0].raw === undefined) {
        styles = rawStyles
      } else {
        styles = interleave(rawStyles)
      }

      // do we really want to use the same infra as the web since it only really uses theming?
      let Styled = withCSSContext((props, context) => {
        let mergedProps = pickAssign(testAlwaysTrue, {}, props, {
          theme: props.theme || context.theme
        })
        let stylesWithStyleProp = styles
        if (props.style) {
          stylesWithStyleProp = styles.concat(props.style)
        }
        const emotionStyles = css.apply(mergedProps, stylesWithStyleProp)

        return React.createElement(
          component,
          pickAssign(pickTest, {}, props, {
            ref: props.innerRef,
            style: emotionStyles
          })
        )
      })
      // $FlowFixMe
      Styled.withComponent = (newComponent: React.ElementType) =>
        createEmotion(newComponent)(...styles)

      Styled.displayName = `emotion(${getDisplayName(component)})`

      return Styled
    }
  }
}

const getDisplayName = primitive =>
  typeof primitive === 'string'
    ? primitive
    : primitive.displayName || primitive.name || 'Styled'
