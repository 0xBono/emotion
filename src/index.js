// @flow
import forEach from '@arr/foreach'
import { StyleSheet } from './sheet'
import { hashString as hash, hashArray, hashObject } from './hash'
import { createMarkupForStyles } from './glamor/CSSPropertyOperations'
import clean from './glamor/clean.js'

export const sheet = new StyleSheet()
// üöÄ
sheet.inject()

export let inserted: { [string]: boolean | void } = {}

type inputVar = string | number

type vars = Array<inputVar>

export function flush () {
  sheet.flush()
  inserted = {}
  sheet.inject()
}

// a simple cache to store generated obj styles
let registered = (sheet.registered = {})

function register (spec) {
  if (!registered[spec.id]) {
    registered[spec.id] = spec
  }
}

function _getRegistered (rule) {
  if (isLikeRule(rule)) {
    let ret = registered[idFor(rule)]
    if (ret == null) {
      throw new Error(
        '[emotion] an unexpected rule cache miss occurred. This is probably a sign of multiple glamor instances in your app. See https://github.com/threepointone/glamor/issues/79'
      )
    }
    return ret
  }
  return rule
}

// The idea on how to merge object class names come from glamorous
// üíÑ
// https://github.com/paypal/glamorous/blob/master/src/get-glamor-classname.js
function getEmotionStylesFromClassName (className) {
  const id = className.trim().slice('css-'.length)
  if (sheet.registered[id]) {
    return sheet.registered[id].style
  } else {
    return []
  }
}

function buildStyles (objs) {
  let computedClassName = ''
  let objectStyles = []

  // This needs to be moved into the core
  forEach(objs, (cls): void => {
    if (typeof cls === 'string') {
      if (cls.trim().indexOf('css-') === 0) {
        objectStyles.push(getEmotionStylesFromClassName(cls))
      } else {
        computedClassName && (computedClassName += ' ')
        computedClassName += cls
      }
    } else {
      objectStyles.push(
        cls
      )
    }
  })

  return { computedClassName, objectStyles }
}

export function css (objs: any, vars: Array<any>, content: () => Array<any>) {
  if (!Array.isArray(objs)) {
    objs = [objs]
  }

  let { computedClassName = '', objectStyles = [] } = buildStyles(content ? objs.concat(content.apply(null, vars)) : objs)
  if (objectStyles.length) {
    computedClassName += ' ' + objStyle.apply(null, objectStyles).toString()
  }

  return computedClassName.trim()
}

export function injectGlobal (src: string[]) {
  const hash = hashArray(src)
  if (!inserted[hash]) {
    inserted[hash] = true
    forEach(src, r => sheet.insert(r))
  }
}

export const fontFace = injectGlobal

export function keyframes (kfm: string, src: string[]) {
  const hash = hashArray(src)
  const animationName = `${kfm}-${hash}`
  if (!inserted[hash]) {
    inserted[hash] = true
    forEach(src, r => sheet.insert(`@keyframes ${animationName} ${r}`))
  }
  return animationName
}

export function hydrate (ids: string[]) {
  forEach(ids, id => (inserted[id] = true))
}

// üç©
// https://github.com/jxnblk/cxs/blob/master/src/monolithic/index.js
type EmotionRule = { [string]: any }

type CSSRuleList = Array<EmotionRule>

type EmotionClassName = {
  [string]: any
}

let cachedCss: (rules: CSSRuleList) => EmotionClassName = typeof WeakMap !==
  'undefined'
  ? multiIndexCache(_css)
  : _css

export function objStyle (...rules: CSSRuleList): EmotionClassName {
  rules = clean(rules)
  if (!rules) {
    return nullrule
  }

  return cachedCss(rules)
}

function _css (rules) {
  let style = {}
  build(style, { src: rules }) // mutative! but worth it.

  let spec = {
    id: hashObject(style),
    style,
    type: 'css'
  }
  return toRule(spec)
}

// define some constants

const isDev = process.env.NODE_ENV === 'development' || !process.env.NODE_ENV
const isTest = process.env.NODE_ENV === 'test'

// takes a string, converts to lowercase, strips out nonalphanumeric.
function simple (str) {
  return str.toLowerCase().replace(/[^a-z0-9]/g, '')
}

// of shape { 'data-css-<id>': '' }
export function isLikeRule (rule: EmotionRule) {
  let keys = Object.keys(rule).filter(x => x !== 'toString')
  if (keys.length !== 1) {
    return false
  }
  return !!/css\-obj\-([a-zA-Z0-9]+)/.exec(keys[0])
}

// extracts id from a { 'css-<id>': ''} like object
export function idFor (rule: EmotionRule) {
  let keys = Object.keys(rule).filter(x => x !== 'toString')
  if (keys.length !== 1) throw new Error('not a rule')
  let regex = /css\-obj\-([a-zA-Z0-9]+)/
  let match = regex.exec(keys[0])
  if (!match) throw new Error('not a rule')
  return match[1]
}

function selector (id: string, path: string = '') {
  if (!id) {
    return path.replace(/\&/g, '')
  }
  if (!path) return `.css-${id}`

  let x = path
    .split(',')
    .map(
      x =>
        x.indexOf('&') >= 0
          ? x.replace(/\&/gm, `.css-${id}`)
          : `.css-${id}${x}`
    )
    .join(',')

  return x
}

function deconstruct (style) {
  // we can be sure it's not infinitely nested here
  let plain, selects, medias, supports
  Object.keys(style).forEach(key => {
    if (key.indexOf('&') >= 0) {
      selects = selects || {}
      selects[key] = style[key]
    } else if (key.indexOf('@media') === 0) {
      medias = medias || {}
      medias[key] = deconstruct(style[key])
    } else if (key.indexOf('@supports') === 0) {
      supports = supports || {}
      supports[key] = deconstruct(style[key])
    } else {
      plain = plain || {}
      plain[key] = style[key]
    }
  })
  return { plain, selects, medias, supports }
}

function deconstructedStyleToCSS (id, style) {
  let { plain, selects, medias, supports } = style
  let css = []

  if (plain) {
    css.push(`${selector(id)}{${createMarkupForStyles(plain)}}`)
  }
  if (selects) {
    Object.keys(selects).forEach((key: string) =>
      css.push(`${selector(id, key)}{${createMarkupForStyles(selects[key])}}`)
    )
  }
  if (medias) {
    Object.keys(medias).forEach(key =>
      css.push(`${key}{${deconstructedStyleToCSS(id, medias[key]).join('')}}`)
    )
  }
  if (supports) {
    Object.keys(supports).forEach(key =>
      css.push(`${key}{${deconstructedStyleToCSS(id, supports[key]).join('')}}`)
    )
  }
  return css
}

// and helpers to insert rules into said sheet
function insert (spec) {
  if (!inserted[spec.id]) {
    inserted[spec.id] = true
    let deconstructed = deconstruct(spec.style)
    deconstructedStyleToCSS(spec.id, deconstructed).map(cssRule =>
      sheet.insert(cssRule)
    )
  }
}

// todo - perf
let ruleCache = {}

function toRule (spec) {
  register(spec)
  insert(spec)
  console.log(spec)
  if (ruleCache[spec.id]) {
    return ruleCache[spec.id]
  }

  let ret = { [`css-${spec.id}`]: '' }
  Object.defineProperty(ret, 'toString', {
    enumerable: false,
    value () {
      return 'css-' + spec.id
    }
  })
  ruleCache[spec.id] = ret
  return ret
}

function log () {
  // eslint-disable-line no-unused-vars
  console.log(this) // eslint-disable-line no-console
  return this
}

function isSelector (key) {
  let possibles = [':', '.', '[', '>', ' '],
    found = false,
    ch = key.charAt(0)
  for (let i = 0; i < possibles.length; i++) {
    if (ch === possibles[i]) {
      found = true
      break
    }
  }
  return found || key.indexOf('&') >= 0
}

function joinSelectors (a, b) {
  let as = a.split(',').map(a => (!(a.indexOf('&') >= 0) ? '&' + a : a))
  let bs = b.split(',').map(b => (!(b.indexOf('&') >= 0) ? '&' + b : b))

  return bs
    .reduce((arr, b) => arr.concat(as.map(a => b.replace(/\&/g, a))), [])
    .join(',')
}

function joinMediaQueries (a, b) {
  return a ? `@media ${a.substring(6)} and ${b.substring(6)}` : b
}

function isMediaQuery (key) {
  return key.indexOf('@media') === 0
}

function isSupports (key) {
  return key.indexOf('@supports') === 0
}

function joinSupports (a, b) {
  return a ? `@supports ${a.substring(9)} and ${b.substring(9)}` : b
}

// flatten a nested array
function flatten (inArr) {
  let arr = []
  for (let i = 0; i < inArr.length; i++) {
    if (Array.isArray(inArr[i])) arr = arr.concat(flatten(inArr[i]))
    else arr = arr.concat(inArr[i])
  }
  return arr
}

// mutable! modifies dest.
function build (dest, { selector = '', mq = '', supp = '', src = {} }) {
  if (!Array.isArray(src)) {
    src = [src]
  }
  src = flatten(src)

  src.forEach(_src => {
    if (isLikeRule(_src)) {
      let reg = _getRegistered(_src)
      if (reg.type !== 'css') {
        throw new Error('cannot merge this rule')
      }
      _src = reg.style
    }
    _src = clean(_src)
    if (_src && _src.composes) {
      console.log('found composes')
      build(dest, { selector, mq, supp, src: _src.composes })
    }
    Object.keys(_src || {}).forEach(key => {
      if (isSelector(key)) {
        if (key === '::placeholder') {
          build(dest, {
            selector: joinSelectors(selector, '::-webkit-input-placeholder'),
            mq,
            supp,
            src: _src[key]
          })
          build(dest, {
            selector: joinSelectors(selector, '::-moz-placeholder'),
            mq,
            supp,
            src: _src[key]
          })
          build(dest, {
            selector: joinSelectors(selector, '::-ms-input-placeholder'),
            mq,
            supp,
            src: _src[key]
          })
        }

        build(dest, {
          selector: joinSelectors(selector, key),
          mq,
          supp,
          src: _src[key]
        })
      } else if (isMediaQuery(key)) {
        build(dest, {
          selector,
          mq: joinMediaQueries(mq, key),
          supp,
          src: _src[key]
        })
      } else if (isSupports(key)) {
        build(dest, {
          selector,
          mq,
          supp: joinSupports(supp, key),
          src: _src[key]
        })
      } else if (key === 'composes') {
        // ignore, we already dealth with it
        console.log('key === composes')
      } else {
        let _dest = dest
        if (supp) {
          _dest[supp] = _dest[supp] || {}
          _dest = _dest[supp]
        }
        if (mq) {
          _dest[mq] = _dest[mq] || {}
          _dest = _dest[mq]
        }
        if (selector) {
          _dest[selector] = _dest[selector] || {}
          _dest = _dest[selector]
        }

        _dest[key] = _src[key]
      }
    })
  })
}

let nullrule: EmotionClassName = {
  // 'data-css-nil': ''
}

Object.defineProperty(nullrule, 'toString', {
  enumerable: false,
  value () {
    return 'css-nil'
  }
})

let inputCaches = typeof WeakMap !== 'undefined'
  ? [nullrule, new WeakMap(), new WeakMap(), new WeakMap()]
  : [nullrule]

let warnedWeakMapError = false

function multiIndexCache (fn) {
  return function (args) {
    if (inputCaches[args.length]) {
      let coi = inputCaches[args.length]
      let ctr = 0
      while (ctr < args.length - 1) {
        if (!coi.has(args[ctr])) {
          coi.set(args[ctr], new WeakMap())
        }
        coi = coi.get(args[ctr])
        ctr++
      }
      if (coi.has(args[args.length - 1])) {
        let ret = coi.get(args[ctr])

        if (registered[ret.toString().substring(4)]) {
          // make sure it hasn't been flushed
          return ret
        }
      }
    }
    let value = fn(args)
    if (inputCaches[args.length]) {
      let ctr = 0,
        coi = inputCaches[args.length]
      while (ctr < args.length - 1) {
        coi = coi.get(args[ctr])
        ctr++
      }
      try {
        coi.set(args[ctr], value)
      } catch (err) {
        if (isDev && !warnedWeakMapError) {
          warnedWeakMapError = true
          console.warn('failed setting the WeakMap cache for args:', ...args) // eslint-disable-line no-console
          console.warn(
            'this should NOT happen, please file a bug on the github repo.'
          ) // eslint-disable-line no-console
        }
      }
    }
    return value
  }
}
