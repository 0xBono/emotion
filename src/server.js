import { sheet, inserted } from './index'

/** ** serverside stuff ****/

// the api's copied from aphrodite, with 1 key difference
// we include *all* the css generated by the app
// to optimize to only include generated styles on the pages
// use renderStaticOptimized
export function renderStatic (fn) {
  let html = fn()
  if (html === undefined) {
    throw new Error('did you forget to return from renderToString?')
  }
  let rules = sheet.sheet.cssRules
  let css = rules.map(r => r.cssText).join('')
  return { html, ids: Object.keys(inserted), css, rules }
}

export function renderStaticOptimized (fn) {
  // parse out ids from html
  // reconstruct css/rules/cache to pass
  let html = fn()
  if (html === undefined) {
    throw new Error('did you forget to return from renderToString?')
  }
  let o = { html, ids: [], css: '', rules: [] }
  let regex = /css-([a-zA-Z0-9]+)/gm
  let match
  let ids = {}
  while ((match = regex.exec(html)) !== null) {
    if (!ids[match[1] + '']) {
      ids[match[1] + ''] = true
    }
  }

  o.rules = sheet.sheet.cssRules.filter(x => {
    let regex = /css-([a-zA-Z0-9]+)/gm
    let match = regex.exec(x.cssText)
    if (match && ids[match[1] + '']) {
      return true
    }
    if (!match) {
      return true
    }
    return false
  })
  o.ids = Object.keys(inserted).filter(id => !!ids[id + ''])
  o.css = o.rules.map(x => x.cssText).join('')

  return o
}
