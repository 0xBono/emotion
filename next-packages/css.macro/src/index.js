// @flow
import { macros } from 'babel-plugin-emotion'

export default macros.css
