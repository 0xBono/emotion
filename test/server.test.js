/**
 * @jest-environment node
*/
/* eslint-env jest */
import React from 'react'
import { renderToString } from 'react-dom/server'
import styled from '../src/styled'
import { renderStatic } from '../src/server'

const color = 'red'

const Main = styled.main`
  display: flex;
`

const Image = styled.img`
  border-radius: 50%;
  height: attr(size px, 50px)
  width: attr(size px, 50px)
  background-color: ${color}
`

const Page = () => <Main><Image size={30} /><Image size={100} /><Image /></Main>

describe('server', () => {
  describe('renderStatic', () => {
    test('returns static css', () => {
      expect(renderStatic(() => renderToString(<Page />))).toMatchSnapshot()
    })
  })
  describe('renderStaticOptimized', () => {
    test('returns static css', () => {
      expect(renderStatic(() => renderToString(<Page />))).toMatchSnapshot()
    })
  })
})
