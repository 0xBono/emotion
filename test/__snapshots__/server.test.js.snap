// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server renderStatic returns static css 1`] = `
Object {
  "css": ".css-Main-ov1ktg-pvm7rn { display: -webkit-box; display: -moz-box; display: -ms-flexbox; display: -webkit-flex; display: flex; }.css-Image-z4ry8g-1qx25bk { border-radius: 50%;
  height: 30;
  width: 30;
  background-color: red }.css-Image-z4ry8g-141jzk4 { border-radius: 50%;
  height: 100;
  width: 100;
  background-color: red }.css-Image-z4ry8g-jpu7qc { border-radius: 50%;
  height: 50pxpx;
  width: 50pxpx;
  background-color: red }",
  "html": "<main class=\\"css-Main-ov1ktg-pvm7rn\\" data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"1383946074\\"><img size=\\"30\\" class=\\"css-Image-z4ry8g-1qx25bk\\" data-reactid=\\"2\\"/><img size=\\"100\\" class=\\"css-Image-z4ry8g-141jzk4\\" data-reactid=\\"3\\"/><img class=\\"css-Image-z4ry8g-jpu7qc\\" data-reactid=\\"4\\"/></main>",
  "ids": Array [
    "pvm7rn",
    "1qx25bk",
    "141jzk4",
    "jpu7qc",
  ],
  "rules": Array [
    Object {
      "cssText": ".css-Main-ov1ktg-pvm7rn { display: -webkit-box; display: -moz-box; display: -ms-flexbox; display: -webkit-flex; display: flex; }",
    },
    Object {
      "cssText": ".css-Image-z4ry8g-1qx25bk { border-radius: 50%;
  height: 30;
  width: 30;
  background-color: red }",
    },
    Object {
      "cssText": ".css-Image-z4ry8g-141jzk4 { border-radius: 50%;
  height: 100;
  width: 100;
  background-color: red }",
    },
    Object {
      "cssText": ".css-Image-z4ry8g-jpu7qc { border-radius: 50%;
  height: 50pxpx;
  width: 50pxpx;
  background-color: red }",
    },
  ],
}
`;

exports[`server renderStaticOptimized returns static css 1`] = `
Object {
  "css": ".css-Main-ov1ktg-pvm7rn { display: -webkit-box; display: -moz-box; display: -ms-flexbox; display: -webkit-flex; display: flex; }.css-Image-z4ry8g-1qx25bk { border-radius: 50%;
  height: 30;
  width: 30;
  background-color: red }.css-Image-z4ry8g-141jzk4 { border-radius: 50%;
  height: 100;
  width: 100;
  background-color: red }.css-Image-z4ry8g-jpu7qc { border-radius: 50%;
  height: 50pxpx;
  width: 50pxpx;
  background-color: red }",
  "html": "<main class=\\"css-Main-ov1ktg-pvm7rn\\" data-reactroot=\\"\\" data-reactid=\\"1\\" data-react-checksum=\\"1383946074\\"><img size=\\"30\\" class=\\"css-Image-z4ry8g-1qx25bk\\" data-reactid=\\"2\\"/><img size=\\"100\\" class=\\"css-Image-z4ry8g-141jzk4\\" data-reactid=\\"3\\"/><img class=\\"css-Image-z4ry8g-jpu7qc\\" data-reactid=\\"4\\"/></main>",
  "ids": Array [
    "pvm7rn",
    "1qx25bk",
    "141jzk4",
    "jpu7qc",
  ],
  "rules": Array [
    Object {
      "cssText": ".css-Main-ov1ktg-pvm7rn { display: -webkit-box; display: -moz-box; display: -ms-flexbox; display: -webkit-flex; display: flex; }",
    },
    Object {
      "cssText": ".css-Image-z4ry8g-1qx25bk { border-radius: 50%;
  height: 30;
  width: 30;
  background-color: red }",
    },
    Object {
      "cssText": ".css-Image-z4ry8g-141jzk4 { border-radius: 50%;
  height: 100;
  width: 100;
  background-color: red }",
    },
    Object {
      "cssText": ".css-Image-z4ry8g-jpu7qc { border-radius: 50%;
  height: 50pxpx;
  width: 50pxpx;
  background-color: red }",
    },
  ],
}
`;
